### ISSUE TITLE
Ground Plane Positioning Bug

### ISSUE THREAD

**Issue by fraktality** [Dreadbot/Recycle-Rush#1]
Accurate position and velocity tracking is strictly dependent on the accelerometer being perfectly parallel with the ground.  Any grade will make the robot interpret gravity as lateral acceleration.  This is problematic as the robot will think that it's decelerating when it goes over a bump, thus making it unnecessarily accelerate to compensate.  This skews velocity readings until it completes transit over the bump. Position readings are permanently affected.

In other words, we'll need a straight up IMU if we want to go over the scoring platform more than once.

**Comment by SteveKing** [comment]
On 2015-01-29 08:40 , Parker Stebbins wrote:

> Accurate position and velocity tracking is strictly dependent on the 
> accelerometer being perfectly parallel with the ground. Any grade will 
> make the robot interpret gravity as lateral acceleration. This is 
> problematic as the robot will think that it's decelerating when it 
> goes over a bump, thus making it unnecessarily accelerate to 
> compensate. This skews velocity readings until it completes transit 
> over the bump. Position readings are permanently affected.
> 
> In other words, we'll need a straight up IMU if we want to go over the 
> scoring platform more than once.

Which, I would say, is a very good argument against trying to calculate 
absolute field position based on inertial navigation only.  And I still 
think that even if you could compensate for orientation the 
accelerometers we have aren't going to be sensitive enough.  But go 
ahead and prototype it by assuming a level surface. We'll push the 
roborio around the shop on a cart to see how much error accumulates over 
the time of a match. That'll tell us whether or not it's even feasible 
and worth taking time to consider a way to determine orientation.

## 

Steve King
steve@narbat.com mailto:steve@narbat.com

**Comment by C7C8** [comment]
What if we pointed a camera (or other small imaging device) directly at the floor such that the camera was perpendicular to the floor? The camera takes snapshots every iteration, and compares the current snapshot to the previous snapshot to determine how much the robot has moved. If this sounds impractical, trust me; it's not - it's the same technology that optical mice use, and affords high level of precision. Comparing one image to another and determining how far one is shifted off from the other isn't overly difficult, so that shouldn't cause too many problems.

Or we could just pick up a cheap wireless mouse, modify it if necessary, mount it to the robot, and hook up its wireless receiver to the roboRIO's USB port and see if we can get it to read input from the mouse (spoiler alert: you can, and I know a great library for doing that, one that I've been using for many months now). We could use the mouse input to determine location, and disregard vision positioning systems entirely, assuming we know the relative location of the tote we want to pick up and the zone we want to drive to. This is starting to look like a more practical idea now.

**Comment by SteveKing** [comment]
On 2015-02-03 11:33 , Sourec wrote:

> What if we pointed a camera (or other small imaging device) directly 
> at the floor such that the camera was perpendicular to the floor? The 
> camera takes snapshots every iteration, and compares the current 
> snapshot to the previous snapshot to determine how much the robot has 
> moved. If this sounds impractical, trust me; it's not - it's the same 
> technology that optical mice use, and affords high level of precision. 
> Comparing one image to another and determining how far one is shifted 
> off from the other isn't overly difficult, so that shouldn't cause too 
> many problems.
> 
> Or we could just pick up a cheap wireless mouse, modify it if 
> necessary, mount it to the robot, and hook up its wireless receiver to 
> the roboRIO's USB port and see if we can get it to read input from the 
> mouse (spoiler alert: you can, and I know a great library for doing 
> that, one that I've been using for many months now). We could use the 
> mouse input to determine location, and disregard vision positioning 
> systems entirely, assuming we know the relative location of the tote 
> we want to pick up and the zone we want to drive to. This is starting 
> to look like a more practical idea now.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/Dreadbot/Dreadbot-V/issues/1#issuecomment-72683348.

Potential problems: Can we process images quickly enough to keep 
sufficient "landmarks" in view in sequential frames? This is 
particularly a problem with using a generic camera and doing image 
processing in software, but could be a problem with a mouse as well. The 
robot will be moving faster than someone typically can push a mouse 
across the desk.

Specific to using a mouse, it would have to be pretty much on the 
ground. Their focal length is set up to be almost in contact with the 
surface. Try lifting a mouse a quarter inch off the desk and moving it 
around. If it responds at all it's going to jitter all over the place.

But hey, I'm naturally a pessimist when it comes to these things. If you 
think it can be done, put a prototype together and let's see!

PS: A wired mouse would be better. More reliable, no additional battery, 
and less concern about causing or receiving interference. I'm not even 
sure a wireless device is legal on the robot. We can make the cable as 
long as we need it, and secure it to whatever's attaching the mouse to 
the bot so it's not in the way.

## 

Steve King
steve@narbat.com mailto:steve@narbat.com

**Comment by C7C8** [comment]
I should be able to determine, using a basic timing setup, how long every "tick"/update is during autonomous. Multiply that time by the highest velocity of the robot in M/S and that should give me the maximum distance the robot moves during the time difference. If it's smaller than the the maximum width of the floorspace a camera can view at a time, there shouldn't be too many difficulties. For a mouse, since all mice seem to be different, we'd have to test it at high speed (that sounds fun!).

For the issue with distancing a mouse above the ground, this was something I considered as well. I did some tests on it, and found that a mouse can get some level of accuracy. With a more precise way of holding the mouse, it might be possible to gain the level of accuracy needed. Of course, it might also be possible for the engineering/design group to rig some setup by which the mouse is literally on the floor, which would end up fixing that particular problem.
